agent/lib_monolog.c

```diff
@@ -0,0 +1,416 @@
+
+  return ZEND_HASH_APPLY_KEEP;
+}
+
+/*
+ * Purpose : Iterate over $context array and format each element
+ *
+ * Params  : string buffer to store formatted output and
+ * Monolog\Logger::addRecord argument list
+ *
+ * Returns : A new string with Monolog's log context
+ */
+static char* nr_monolog_fmt_context(char* strbuf,
+                                    HashTable* context TSRMLS_DC) {
+  strbuf = nr_str_append(strbuf, "[", "");
+
+  nr_php_zend_hash_zval_apply(context,
+                              (nr_php_zval_apply_t)nr_monolog_fmt_context_item,
+                              (void*)&strbuf TSRMLS_CC);
+
+  return nr_str_append(strbuf, "]", "");
+}
+
+/*
+ * Purpose : Convert $context argument of Monolog\Logger::addRecord to a string
+ *
+ * Params  : # of Monolog\Logger::addRecord arguments, and
+ * Monolog\Logger::addRecord argument list
+ *
+ * Returns : A new string with Monolog's log context
+ */
+static char* nr_monolog_get_context(const size_t argc,
+                                    NR_EXECUTE_PROTO TSRMLS_DC) {
+  char* context = nr_strdup("");
+  zval* context_arg = NULL;
+
+  if (3 > argc) {
+    nrl_verbosedebug(NRL_INSTRUMENT, "%s: $context not available", __func__);
+    goto return_context;
+  }
+
+  context_arg = nr_php_arg_get(3, NR_EXECUTE_ORIG_ARGS TSRMLS_CC);
+
+  if (NULL == context_arg) {
+    nrl_verbosedebug(NRL_INSTRUMENT,
+                     "%s: $context not defined, unable to get log context",
+                     __func__);
+    goto return_context;
+  } else if (!nr_php_is_zval_valid_array(context_arg)) {
+    nrl_verbosedebug(NRL_INSTRUMENT,
+                     "%s: expected $context be a valid array, got type %d",
+                     __func__, Z_TYPE_P(context_arg));
+    goto return_context;
+  }
+
+  if (0 == nr_php_zend_hash_num_elements(Z_ARRVAL_P(context_arg))) {
+    nrl_verbosedebug(NRL_INSTRUMENT, "%s: $context has no elements", __func__);
+    goto return_context;
+  }
+
+  context = nr_monolog_fmt_context(context, Z_ARRVAL_P(context_arg) TSRMLS_CC);
+
+return_context:
+  nr_php_arg_release(&context_arg);
+  return context;
+}
+
+/*
+ * Purpose : Combine $message and $context arguments of
+ * Monolog\Logger::addRecord into a single string to be used as a message
+ * property of the log event.
+ *
+ * Params  : # of Monolog\Logger::addRecord arguments, and
+ * Monolog\Logger::addRecord argument list
+ *
+ * Returns : A new string with a log record message; caller must free
+ */
+static char* nr_monolog_build_message(const size_t argc,
+                                      NR_EXECUTE_PROTO TSRMLS_DC) {
+  char* message_and_context = nr_strdup("");
+
+  char* message = nr_monolog_get_message(NR_EXECUTE_ORIG_ARGS TSRMLS_CC);
+  message_and_context = nr_str_append(message_and_context, message, "");
+  nr_free(message);
+
+  char* context = nr_monolog_get_context(argc, NR_EXECUTE_ORIG_ARGS TSRMLS_CC);
+  message_and_context = nr_str_append(message_and_context, context, " ");
```

Comment:
```suggestion
    if (0 < nr_strlen(context))
        message_and_context = nr_str_append(message_and_context, context, " ");
```

This avoids appending a blank space on every message which has an empty context.

Reply:
@mfulb Thanks for great suggestion! I addressed this in bcbcaefad9e1380eb03bcf9c86e292090328cb63. However I opted to use `nr_strempty` instead of `nr_strlen` as the former function exists for this particular use case.